src/algo/articulation_points.rs

```diff
@@ -0,0 +1,146 @@
+use crate::visit;
+use crate::visit::{EdgeRef, IntoEdges, IntoNodeReferences, NodeIndexable, NodeRef};
+use std::cmp::min;
+use std::collections::{HashMap, HashSet};
+use std::hash::Hash;
+
+/// \[Generic\] Find articulation points in a graph using [Tarjan's algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm).
+///
+/// Compute the articulation points of a graph (Nodes, which would increase the number of connected components in the graph.
+///
+/// # Arguments
+/// * `graph`: A directed graph
+///
+/// # Returns
+/// * `HashSet`: Hashset of the node ids which correspond to the articulation points of the grpah.
+///
+/// # Examples
+/// ```rust
+/// use petgraph::{
+///     algo::articulation_points,
+///     graph::{NodeIndex, UnGraph},
+///     algo::articulation_points::articulation_points,
+/// };
+///
+/// fn main() {
+///     let mut gr = UnGraph::<&str, ()>::new_undirected();
+///     let a = gr.add_node("A");
+///     let b = gr.add_node("B");
+///     let c = gr.add_node("C");
+///
+///     gr.add_edge(a, b, ());
+///     gr.add_edge(b, c, ());
+///
+///     let articulation_points: Vec<&str> = articulation_points(&gr)
+///         .into_iter()
+///         .map(|node_idx| gr[node_idx])
+///         .collect();
+///
+///     // Articulation Points: ["B"]
+///     println!("Articulation Points: {:?}", articulation_points);
+/// }
+/// ```
+pub fn articulation_points<G>(g: G) -> HashSet<G::NodeId>
+where
+    G: IntoNodeReferences + IntoEdges + NodeIndexable + visit::GraphProp,
+    G::NodeWeight: Clone,
+    G::EdgeWeight: Clone + PartialOrd,
+    G::NodeId: Eq + Hash,
+{
+    let graph_size = g.node_references().size_hint().0;
+
+    let mut visited = vec![false; graph_size];
+    let mut disc = vec![usize::MAX; graph_size];
+    let mut low = vec![usize::MAX; graph_size];
+    let mut parent = vec![usize::MAX; graph_size];
+    let mut ap = HashSet::with_capacity(graph_size);
+    let mut time = 0;
+
+    for node in g.node_references() {
+        let node_id = g.to_index(node.id());
+        if !visited[node_id] {
+            _dfs(
+                &g,
+                node_id,
+                &mut visited,
```

Comment:
nitpick: This could be an auxiliary state struct.

Reply:
Added a struct `ArticulationPointTracker`, I am kindof unhappy with the naming tho.

Reply:
I'd normally call it "Context", but it's private so it doesn't really matter.